Add option to restrict permissions to group types

OskarPersson opened the pull request:
This allows us to specify which types of groups related to the base group should get the different permissions

```
custom_permissions = {
    'groups_downstream': {'developer': ['change', 'delete'], 'default': ['view']},
}
```

vittoriozamboni commented:
Hi Oskar, thanks for the PR. 
I am looking into the code and seems useful and well written. Do you mind to add this new functionality also to the documentation?
From my understanding, this is to allow permissions for related groups (upstream, downstream, siblings) to be specified as dictionary where the key is the group name and the value is the list of permissions (in addition to the default behaviour of having only a list). Is this correct?

OskarPersson commented:
I added the functionality to the documentation and updated the tests to assign the object directly to the group to avoid having to assign the members to the `company` group.

Allan-Nava commented:
So if i create a "**manager**" group to manage a Team(group), they(manager) can modify only for the team group inter?

@OskarPersson

OskarPersson commented:
@Allan-Nava In the following example, the child groups of `company` that are of type `leader` or `referent` would get `change` and `delete` permissions. All child groups of `company` would get `view` permission. All parent groups of `company` would get `view` permission

```
custom_permissions = {
    'owner': ['view', 'change', 'delete',],
    'group': ['view', 'change',],
    'groups_upstream': ['view'],
    'groups_downstream': {
        'leader': ['change', 'delete'],
        'referent': ['change', 'delete'],
        'default': ['view']
    },
    'groups_siblings': ['view'],
}
john.assign_object(company, site, custom_permissions=custom_permissions)
```

Allan-Nava commented:
So @OskarPersson , can I extend the standard settings.py inside **groups_manager** app ? 

Like this:
```
PERMISSIONS = {
    'owner': ['view', 'change', 'delete',],
    'group': ['view', 'change',],
    'groups_upstream': ['view'],
    #'groups_downstream': ,
    'groups_downstream': {
        'leader': ['change', 'delete'], 
        'referent': ['change', 'delete'], 
        'default': ['view']
    },
    'groups_siblings': ['view'],
}
```

I need a general permission that a leader of specif group can delete or change the belong group.

Sorry for my bad english

OskarPersson commented:
@Allan-Nava Yes, in your settings.py you could have the following to extend the default permissions:

```
GROUPS_MANAGER = {
    'PERMISSIONS': {
        'owner': ['view', 'change', 'delete',],
        'group': ['view', 'change',],
        'groups_upstream': ['view'],
        #'groups_downstream': ,
        'groups_downstream': {
            'leader': ['change', 'delete'], 
            'referent': ['change', 'delete'], 
            'default': ['view']
        },
        'groups_siblings': ['view'],
    }
}
```

Allan-Nava commented:
Ok, but I have to assign the object programmatically or can I make it standard for those who belong to the leading group ?

`john.assign_object(company, site, custom_permissions=custom_permissions)`

Because I need in a "football clubs" the Leader and Referent can handle the players inside the belong group.

Thanks for support 💯

OskarPersson commented:
@Allan-Nava You have to programmatically assign the object to at least one user or group in order to get permissions, yes.

Allan-Nava commented:
@OskarPersson Ok, but if I want to associate "john" in adminPanel with leader privilges how can I do?

I have done with user/groups like user management as a spotify. 

![screen shot 2017-11-28 at 09 40 20](https://user-images.githubusercontent.com/22498435/33309946-4472a61a-d420-11e7-8fee-f38b3e7710d4.png)
![screen shot 2017-11-28 at 09 57 16](https://user-images.githubusercontent.com/22498435/33310648-9a3cee0a-d422-11e7-9492-6620dc6cd985.png)

OskarPersson commented:
@Allan-Nava I'm not sure I understand what you mean. Also, this is probably not the right place for discussing it neither. Maybe create a detailed question on stackoverflow or similar?

Allan-Nava commented:
@OskarPersson Why is not the right place? 

I added the functionality of object permission of django guardian in the adminPanel, but I need to implement to group.. Any suggestion?

Thanks

Allan-Nava commented:
I tried to assign object via code but doesn't work, @OskarPersson 

![screen shot 2017-12-15 at 11 23 19](https://user-images.githubusercontent.com/22498435/34037815-727db672-e18a-11e7-9db8-7e9673841b9b.png)

I have an istance of member and Group juve, I want to assign the default permission
